### PR TITLE
[WIP] Adds a Helm check for a new version of Helm

### PR DESCRIPTION
**Pull request by mattfarina**
There are a few things this needs to do before it can be merged.

1. Get a real releases URL and a process documented to update it
2. Question, should the notification happen on first run?
3. Does the opt-out work ok?

The design of this was so that it could be leveraged from within another client that pulls helm in and with a helper to make it easy for helm operating locally. Is the design layout ok?

### CODE REVIEW

**Comment by mattfarina** [comment]
FYI, CircleCI is passing for this... https://circleci.com/gh/kubernetes/helm/tree/pull%2F4353

File: pkg/helm/environment/environment.go [hunk 1]
```diff
@@ -59,6 +61,7 @@ func (s *EnvSettings) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&s.Debug, "debug", false, "enable verbose output")
 	fs.StringVar(&s.TillerNamespace, "tiller-namespace", "kube-system", "namespace of Tiller")
 	fs.Int64Var(&s.TillerConnectionTimeout, "tiller-connection-timeout", int64(300), "the duration (in seconds) Helm will wait to establish a connection to tiller")
+	fs.BoolVar(&s.SkipSelfUpdateCheck, "skip-self-update-check", false, "disable checking for updates to Helm")
```
Comment (bacongobbler) [inline pkg/helm/environment/environment.go]: double negatives always throw me off. What do you think about re-naming this to `check-for-updates` or something, and changing the default to true?

File: pkg/helm/notify/notify.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/Masterminds/semver"
+	"k8s.io/helm/pkg/version"
+)
+
+var (
```
Comment (bacongobbler) [inline pkg/helm/notify/notify.go]: do we wanna make this a const instead if it's a private variable? I don't see anywhere in the package where we are going to change the time layout.

**Comment by bacongobbler** [comment]
> Get a real releases URL and a process documented to update it

Hmm, assuming we publish this file via the [existing release process](https://github.com/helm/helm/blob/f8278afcce2e6a1380108e768aa75cfdbf6ad48c/.circleci/deploy.sh#L63-L64), we could just point to the cloud storage endpoint hosting the Helm releases. However, if CI's gated to only upload on official releases, that would mean that there's a small point in time between the latest release and when this file is updated. WDYT?

> Question, should the notification happen on first run?

I think so, yes!

> Does the opt-out work ok?

Is there a way to persist the opt-out check? From a cursory glance it appears that the user has to constantly supply the feature flag to disable it. Should we check for the presence of a file in $HELM_HOME?

File: pkg/helm/environment/environment.go (hunk 1 above)
Comment (mattfarina) [reply pkg/helm/environment/environment.go]: How would someone set it to false at the command line? How would the internal processing map the CLI value to the internal variable?

Do you have a good example of making it a positive?

File: pkg/helm/notify/notify.go (hunk 2 above)
Comment (mattfarina) [reply pkg/helm/notify/notify.go]: Now a bad idea. Will do that in next update.

**Comment by mattfarina** [comment]
> Hmm, assuming we publish this file via the existing release process, we could just point to the cloud storage endpoint hosting the Helm releases.

When it comes to Helm 3 are we going to host releases in the same place? That is, a Google Cloud object storage container with Google in the name rather than some DNS name we've put in front of that.

I would prefer the update URL be hosted behind the helm.sh DNS somewhere so we can change the implementation later.

> However, if CI's gated to only upload on official releases, that would mean that there's a small point in time between the latest release and when this file is updated. WDYT?

I would keep the current CI process. Update this separately. Then we can control when the release goes out. For example, we might tag a release and have it generated but have a window while we finish the release notes before telling people. We can choose when we tell people.

**Comment by mattfarina** [comment]
> Is there a way to persist the opt-out check?

I can add something

I was also thinking of adding a command to check

File: pkg/helm/environment/environment.go (hunk 1 above)
Comment (bacongobbler) [reply pkg/helm/environment/environment.go]: Yeah! The variable name could change to `s.PerformSelfUpdateCheck`, which is set to `true` by default.

Then, on line 98 in helm.go we'd change `if !settings.SkipSelfUpdateCheck` to `if settings.PerformSelfUpdateCheck`. The `!settings.SkipSelfUpdateCheck` to `if settings.PerformSelfUpdateCheck` is the double negative I'm referencing if it wasn't totally clear btw. Does that help make sense or do you wanna chat about this offline?

> How would someone set it to false at the command line?

`helm --check-for-updates=false`

**Comment by technosophos** [comment]
It seems like the best place to fetch release info is from the releases page on GitHub. I know there is an API endpoint for releases. Could we use that?

I am fairly certain that we will be moving the releases off of the current GCS location simply because that is part of the Kubernetes project.

**Comment by bacongobbler** [comment]
> It seems like the best place to fetch release info is from the releases page on GitHub. I know there is an API endpoint for releases. Could we use that?

ref: https://github.com/technosophos/helm-template/blob/c96a3c60ccc2b44e64aaddb43edb09a8802948bc/install-binary.sh#L66-L75

**Comment by technosophos** [comment]
That snippet actually was borrowed from the `get` script in this repo, which does approximately the same thing.

And I just realized that this is a 100% Matts thread.

**Comment by mattfarina** [comment]
@technosophos There were a few reasons I didn't want to go with the GitHub releases. Curious to know what you think of them...

1. You'll notice in the release data is the checksums. Since we don't control the releases on GitHub there is no way to inject checksums. I had a tangential idea to host the checksums in a separate location from the downloads.
1. Anything we release, whether pre-release or alpha, is in the release list. I wasn't thinking that would be in the update information for which we check against.
1. Decoupling notifications from releases. For example, some projects do notifications the day after a release.
1. Portability... what if we someday want to move releases elsewhere?
1. The GitHub API v3 has a rate limit of 60 per hour when anonymous (graphql v4 api cannot be hit anonymously). We could run into problems within CI systems like TravisCI. I've experienced this in the past.

So, this is why I was thinking of doing it outside of GitHub releases. Thoughts?

**Comment by technosophos** [comment]
Oh, I was reading the patch as using GitHub for release info anyway. Yeah, I can see how actually having a dedicated URL with checksums will make things a little more secure (in the sense that the person building it can generate the checksum). So I am okay with that. Where, though, would we put this information? `helm.sh`?

**Comment by bacongobbler** [comment]
ping. As per [the announcement on the mailing list](https://lists.cncf.io/g/cncf-helm/topic/last_v2_feature_release/34301238), Helm 2.15.0 is the last release where we will be merging new features for Helm 2. Are you still interested in getting this one in before the 2.15 release, @mattfarina?

If you don't have the time right now but you're still interested in pushing this along the finish line, would you mind re-targeting this against Helm 3?

Thanks!

**Comment by mattfarina** [comment]
Closing as it needs massive rework before going in. If someone else what's to do this they should feel free to. Otherwise, I'll rework this in a future PR.

**Pull request closed without merging**